### PR TITLE
Implement pipeline pattern

### PR DESCRIPTION
**Pull request by dennismeister93**
Signed-off-by: Dennis Meister <dennis.meister@bosch.com>

Implement pipeline pattern for extracting needed code snippet blocks

### CODE REVIEW

File: src/code-formatter.ts [hunk 1]
```diff
@@ -13,23 +13,47 @@
...
+    codeSnippetForTemplate: string = '';
+}
 export class CodeFormatter {
-    private basicImportsArray: string[] = [];
-    private variablesArray: string[][] = [];
-    private variableNames: string[] = [];
-    private seperateClassesArray: string[][] = [];
-    private seperateMethodsArray: string[][] = [];
-    private codeSnippetStringArray: string[] = [];
+    private codeContext: CodeContext = {
```
Comment (doosuu) [inline src/code-formatter.ts]: Is this explicit initialization needed? We already have empty defaults in the class.

File: src/code-formatter.ts [hunk 2]
```diff
@@ -13,23 +13,47 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import { Buffer } from 'buffer';
-import { DIGITAL_AUTO, INDENTATION, PYTHON, VELOCITAS } from './utils/codeConstants';
+import { ExtractClassesStep } from './pipeline/extract-classes';
```
Comment (doosuu) [inline src/code-formatter.ts]: Finding shall be a file wide finding, ignore the line it is attached to:
`code-foramtter.ts`: File is not a formatter, it is a converter

File: src/code-formatter.ts [hunk 3]
```diff
@@ -64,333 +88,62 @@ export class CodeFormatter {
         return indentedCodeSnippet;
     }
 
-    private adaptCodeSnippet(base64CodeSnippet: string): string {
+    private adaptCodeSnippet(base64CodeSnippet: string): void {
         let decodedBase64CodeSnippet = Buffer.from(base64CodeSnippet, CONTENT_ENCODINGS.base64 as BufferEncoding).toString(
```
Comment (doosuu) [inline src/code-formatter.ts]: Decoding of the base64 string should be done outside of the method. It is a separate concern.

File: src/code-formatter.ts [hunk 4]
```diff
@@ -64,333 +90,63 @@ export class CodeFormatter {
...
-            }
-        }
-        return array;
+        pipeline.forEach((pipelineStep) => pipelineStep.execute(this.codeContext));
     }
 
     private extractMainPyBaseStructure(base64MainPyContentData: string): string {
-        let seperateClassForTemplate = '';
-        let seperateMethodsForTemplate = '';
         let decodedMainPyContentData = Buffer.from(base64MainPyContentData, CONTENT_ENCODINGS.base64 as BufferEncoding).toString(
```
Comment (doosuu) [inline src/code-formatter.ts]: Decoding could be done outside the method, it is a seperate concern

File: src/code-formatter.ts [hunk 5]
```diff
@@ -418,6 +173,9 @@ export class CodeFormatter {
                 (finalCode as string[])[index] = codeLine.replace('",', ': %s",');
             }
         });
+        if (!finalCode.some((line: string) => line.includes(VELOCITAS.CLASS_METHOD_SIGNATURE))) {
+            (finalCode as string[]).splice(finalCode.indexOf(VELOCITAS.IMPORT_DATAPOINT_REPLY), 1);
+        }
         const formattedFinalCode = this.createMultilineStringFromArray(finalCode);
         const encodedNewMainPy = Buffer.from(formattedFinalCode, CONTENT_ENCODINGS.utf8 as BufferEncoding).toString(
```
Comment (doosuu) [inline src/code-formatter.ts]: Encoding should be done outside of the method, it is a seperate concern

File: src/code-formatter.ts [hunk 6]
```diff
@@ -13,23 +13,49 @@
...
+        memberVariables: '',
+        seperateClassesArray: [],
+        seperateClasses: '',
+        seperateMethodsArray: [],
+        seperateMethods: '',
+        codeSnippetStringArray: [],
+        codeSnippetForTemplate: '',
+    };
 
     public formatMainPy(base64MainPyContentData: string, base64CodeSnippet: string, appName: string): string {
```
Comment (doosuu) [inline src/code-formatter.ts]: Add method documentation.

File: src/code-formatter.ts [hunk 7]
```diff
@@ -64,333 +90,63 @@ export class CodeFormatter {
         return indentedCodeSnippet;
     }
 
-    private adaptCodeSnippet(base64CodeSnippet: string): string {
+    private adaptCodeSnippet(base64CodeSnippet: string): void {
```
Comment (doosuu) [inline src/code-formatter.ts]: Ignore the line this finding is attached to. The utility methods above this one like indent, createArray..., createMultiline should be moved to a utils file to have this file as short and concise as possible.

File: src/pipeline/prepare-code-snippet.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import { CodeContext } from '../code-formatter';
+import { DIGITAL_AUTO, PYTHON } from '../utils/codeConstants';
+import { PipelineStep } from './pipeline-base';
+
+export class PrepareCodeSnippetStep extends PipelineStep {
```
Comment (doosuu) [inline src/pipeline/prepare-code-snippet.ts:9]: Add missing class documentation. What does this pipeline step do?

File: src/pipeline/prepare-code-snippet.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        context.codeSnippetStringArray = this.removeSubstringFromArray(context.codeSnippetStringArray, DIGITAL_AUTO.VEHICLE_INIT);
+        context.codeSnippetStringArray = this.removeSubstringFromArray(
+            context.codeSnippetStringArray,
+            PYTHON.IMPORT_DEPENDENCY_FROM,
+            PYTHON.IMPORT
+        );
+        context.codeSnippetStringArray = this.removeSubstringFromArray(context.codeSnippetStringArray, PYTHON.COMMENT);
+    }
+
+    private removeSubstringFromArray(array: string[], substringOne: string, substringTwo?: string): string[] {
```
Comment (doosuu) [inline src/pipeline/prepare-code-snippet.ts]: If this removes a substring from an array why are there 2 substrings possible as parameters? Should this be named removeSubstring**s**FromArray ?

File: src/pipeline/pipeline-base.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                linesToRemove.push(string);
+            });
+        }
+        linesToRemove.forEach((lineToRemove: string) => {
+            if (codeContext.codeSnippetStringArray.indexOf(lineToRemove) >= 0) {
+                codeContext.codeSnippetStringArray?.splice(codeContext.codeSnippetStringArray.indexOf(lineToRemove), 1);
+            }
+        });
+    }
+    indentCodeSnippet(decodedSnippet: string, indentCount: number): string {
```
Comment (doosuu) [inline src/pipeline/pipeline-base.ts]: This method already exists in code-foramter.ts

File: src/pipeline/pipeline-base.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        const indent = ' ';
+        const indentedCodeSnippet = decodedSnippet.replace(REGEX.FIND_EVERY_LINE_START, indent.repeat(indentCount));
+        return indentedCodeSnippet;
+    }
+    createMultilineStringFromArray(array: string[] | string[][]): string {
```
Comment (doosuu) [inline src/pipeline/pipeline-base.ts]: This method already exists in code-foramter.ts

File: src/pipeline/pipeline-base.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import { CodeContext } from '../code-formatter';
+import { VELOCITAS } from '../utils/codeConstants';
+import { REGEX } from '../utils/regex';
+
+export interface IPipelineStep {
+    execute(context: CodeContext): void;
+    cleanUpCodeSnippet(arrayToCleanUp: string[] | string[][], codeContext: CodeContext): void;
+}
+
+export class PipelineStep implements IPipelineStep {
```
Comment (doosuu) [inline src/pipeline/pipeline-base.ts:14]: Why is there a base implementation of the interface? Add a class documentation.

File: src/pipeline/extract-variables.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
+import { CodeContext } from '../code-formatter';
+import { INDENTATION } from '../utils/codeConstants';
+import { PipelineStep } from './pipeline-base';
+
+export class ExtractVariablesStep extends PipelineStep {
```
Comment (doosuu) [inline src/pipeline/extract-variables.ts:10]: Add missing class documentation. What does this pipeline step do?

File: src/pipeline/extract-methods.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
+import { CodeContext } from '../code-formatter';
+import { DIGITAL_AUTO, INDENTATION, PYTHON, VELOCITAS } from '../utils/codeConstants';
+import { PipelineStep } from './pipeline-base';
+
+export class ExtractMethodsStep extends PipelineStep {
```
Comment (doosuu) [inline src/pipeline/extract-methods.ts:10]: Add missing class documentation. What does this pipeline step do?

File: src/pipeline/extract-imports.ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import { CodeContext } from '../code-formatter';
+import { PYTHON, VELOCITAS } from '../utils/codeConstants';
+import { PipelineStep } from './pipeline-base';
+
+export class ExtractImportsStep extends PipelineStep {
```
Comment (doosuu) [inline src/pipeline/extract-imports.ts:9]: Add missing class documentation. What does this pipeline step do?

File: src/pipeline/extract-classes.ts [hunk 16]
```diff
@@ -0,0 +1,43 @@
+import { CodeContext } from '../code-formatter';
+import { PYTHON } from '../utils/codeConstants';
+import { REGEX } from '../utils/regex';
+import { PipelineStep } from './pipeline-base';
+
+export class ExtractClassesStep extends PipelineStep {
```
Comment (doosuu) [inline src/pipeline/extract-classes.ts:11]: Add missing class documentation. What does this pipeline step do?

File: src/pipeline/create-code-snippet.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import { CodeContext } from '../code-formatter';
+import { INDENTATION, VELOCITAS } from '../utils/codeConstants';
+import { PipelineStep } from './pipeline-base';
+
+export class CreateCodeSnippetForTemplateStep extends PipelineStep {
```
Comment (doosuu) [inline src/pipeline/create-code-snippet.ts:10]: Add missing class documentation. What does this pipeline step do?

**Review by doosuu — COMMENTED** [review]
Logger statements are missing f-strings, look at example_output_1.py line 91

**Review by doosuu — CHANGES_REQUESTED** [review]
Please have a look at the findings.

**Comment by dennismeister93** [comment]
> Logger statements are missing f-strings, look at example_output_1.py line 91

Hi @doosuu 
Unfortunately our linter has problems with using this. 
"Use lazy % formatting in logging functionspylint(logging-fstring-interpolation)"
I would propose we do not change it within this PR - we currently also have a linter improvement story in the sprint

**Review by doosuu — APPROVED** [review]
LGTM

**Pull request merged**